python/ml4ir/model/architectures/dnn.py

```diff
@@ -1,20 +1,58 @@
 import tensorflow as tf
 from tensorflow.keras import layers
+from typing import List
 
 
-def get_architecture_op(num_nodes=512):
-    def architecture_op(ranking_features):
-        first_dense = layers.Dense(num_nodes, activation="relu", name="first_dense")(
-            ranking_features
-        )
-        first_dropout = layers.Dropout(rate=0.3, name="first_dropout")(first_dense)
-        final_dense = layers.Dense(64, activation="relu", name="final_dense")(first_dropout)
-        # _ = tf.keras.layers.Dropout(rate=0.1g)(_)
-        scores = layers.Dense(1, name="scores")(final_dense)
+class DNNLayer:
+    DENSE = "dense"
+    BATCH_NORMALIZATION = "batch_norm"
+    DROPOUT = "dropout"
 
-        # Collapse extra dimensions
-        scores = tf.squeeze(scores, axis=-1)
 
-        return scores
+class DNN:
+    def __init__(self, model_config):
+        self.layer_ops: List = self.define_architecture(model_config)
 
-    return architecture_op
+    @staticmethod
+    def _get_dense_layer(**kwargs):
+        return layers.Dense(**kwargs)
+
+    @staticmethod
+    def _get_batch_norm_layer(**kwargs):
+        return layers.BatchNormalization(**kwargs)
+
+    @staticmethod
+    def _get_dropout_layer(**kwargs):
+        return layers.Dropout(**kwargs)
+
+    def define_architecture(self, model_config):
+        layer_ops = list()
+        for layer_info in model_config["layers"]:
+            layer_type = layer_info.pop("type")
+            if layer_type == DNNLayer.DENSE:
+                layer_op = DNN._get_dense_layer(**layer_info)
+            elif layer_type == DNNLayer.BATCH_NORMALIZATION:
+                layer_op = DNN._get_batch_norm_layer(**layer_info)
+            elif layer_type == DNNLayer.DROPOUT:
+                layer_op = DNN._get_dropout_layer(**layer_info)
+            else:
+                raise KeyError("Dense layer type is not supported : {}".format(layer_type))
+
+            layer_ops.append(layer_op)
+
+        return layer_ops
+
+    def get_architecture_op(self):
+        def _architecture_op(ranking_features):
+            _ = ranking_features
```

Comment:
what does this line do?  can we just actually name our variables?

Reply:
I can rename `_` to `layer_input`. Does that make sense?

Reply:
yeah that's better.  Anything is better than "_" ;P

Reply:
although since you're using it as a temp variable, maybe calling it just "layer" is better.  It doesn't make sense to call it anything "input" related after the beginning.

Reply:
It is an input to the layer, no?